staging/src/k8s.io/endpointslice/util/endpoint/controller_utils.go

```diff
@@ -85,7 +85,7 @@ func NewPortMapKey(endpointPorts []discovery.EndpointPort) PortMapKey {
 // DeepHashObjectToString creates a unique hash string from a go object.
 func DeepHashObjectToString(objectToWrite interface{}) string {
```

Comment:
Maybe even flatten `k8s.io/endpointslice/util` into `k8s.io/endpointslice`? Looking at some of the functions like `ShouldPodBeInEndpoints`, I think they belong in the main pkg. See  https://www.kubernetes.dev/docs/guide/coding-convention/#directory-and-file-conventions:

> Avoid general utility packages. Packages called “util” are suspect.

Reply:
yeah, it does not make have much sense have so many utils folders, in theory `k8s.io/endpointslice/util/endpoint` should only be used for the endpoints controller, but I'm not surprised we ended mixing things

Reply:
I think we have to limit ourselves to k8s.io/endpointslice/util, since `k8s.io/endpointslice/topologycache` and `k8s.io/endpointslice/metrics` import utils as well, and if not mistaken, we will end up with cyclic imports

Reply:
It looks like `k8s.io/endpointslice/topologycache` no longer depends on `k8s.io/endpointslice/util`, and only `k8s.io/endpointslice/metrics` does. But now I'm wondering whether `metrics` should be flattened into `k8s.io/endpointslice` as well ..

Is there any value in having metrics be a separate module? It's only really used by the reconciler, it doesn't seem reusable anywhere else. This would let us reduce the API surface substantially.

Reply:
There are a bunch of `/util` functions that are only used within `k8s.io/endpointslice` (incl. `metrics` and `topologycache`) but they're part of the public API. Instead of flattening, maybe a better approach is to move those functions into `/internal` instead?

Reply:
`metrics` is used in the controller e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go#L564